File: misc/docs/HTTP-PDA_Validation/PDAController.cpp
#include "PDAController.hpp"
#include <sstream>

PDAController::PDAController() : traceIndex(0) {}

bool PDAController::loadPacket(const std::string& data) {
    payload = data;
    traceIndex = 0;
    return true;
}

bool PDAController::validate() {
    bool ok = pda.validate(payload);
    traceIndex = 0; // reset stepping index so frontend will iterate from start
    return ok;
}

bool PDAController::hasMoreSteps() {
    return traceIndex < pda.getTrace().size();
}

std::string PDAController::getNextTraceStep() {
    if (!hasMoreSteps()) return "";

    const PDATrace& t = pda.getTrace()[traceIndex++];

    std::ostringstream oss;
    // show state, input (with readable escape for CR/LF), stack top, action
    std::string inputDesc;
    if (t.input == 0) inputDesc = "ε";
    else if (t.input == '\r') inputDesc = "\\r";
    else if (t.input == '\n') inputDesc = "\\n";
    else {
        inputDesc = std::string(1, t.input);
    }

    oss << "State=" << static_cast<int>(t.state)
        << " Input=" << inputDesc
        << " StackTop=" << t.stackTop
        << " Action=" << t.action;

    return oss.str();
}

File: frontend/src/utils/pdaController.ts
// TypeScript PDA Controller wrapper based on docs/HTTP-PDA_Validation/PDAController.cpp

import { PDAEngine, PDATrace } from './pdaEngine';

export class PDAController {
  private pda: PDAEngine;
  private payload: string = '';
  private traceIndex: number = 0;

  constructor() {
    this.pda = new PDAEngine();
    this.traceIndex = 0;
  }

  loadPacket(data: string): boolean {
    this.payload = data;
    this.traceIndex = 0;
    return true;
  }

  validate(): boolean {
    const ok = this.pda.validate(this.payload);
    this.traceIndex = 0; // reset stepping index
    return ok;
  }

  hasMoreSteps(): boolean {
    return this.traceIndex < this.pda.getTrace().length;
  }

  getNextTraceStep(): string {
    if (!this.hasMoreSteps()) return '';

    const trace = this.pda.getTrace();
    const t = trace[this.traceIndex++];

    const inputDesc = t.input;
    return `State=${t.state} Input=${inputDesc} StackTop=${t.stackTop} Action=${t.action}`;
  }

  getAllTrace(): PDATrace[] {
    return this.pda.getTrace();
  }

  getHeaders(): Map<string, string> {
    return this.pda.getHeaders();
  }
}

